[Notifier] Added telnyx notifier bridge

krasilnikovs opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT

Added telnyx notifier bridge for sending sms

https://developers.telnyx.com/docs/api/v2/messaging/Messages?lang=curl#createMessage

krasilnikovs committed:
Added telnyx notifier bridge

jderusse commented on src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php:
```diff
@@ -0,0 +1,114 @@
+/**
+ * @author Mihail Krasilnikov <mihail.krasilnikov.j@gmail.com>
+ *
+ * @experimental in 5.2
```
5.3

krasilnikovs replied:
Thanks, changed to 5.3

jderusse reviewed (commented):
Thank you for this contribution.

Don't forget to add an entry in the ChangeLog. and in UnsupportedSchemeException files

jschaedl commented on src/Symfony/Component/Notifier/Bridge/Telnyx/README.md:
```diff
@@ -0,0 +1,12 @@
+Telnyx Notifier
+=================
```
```suggestion
===============
```

krasilnikovs replied:
Done

jschaedl commented on src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php:
```diff
@@ -0,0 +1,100 @@
+
+    public function __toString(): string
+    {
+        return "telnyx://{$this->apiKey}?from={$this->from}";
```
Using `sprintf` would be more readable imho.

krasilnikovs replied:
In other transports uses `sprintf` and for consistency I decided to use `sprintf` as well.

jschaedl commented on src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php:
```diff
@@ -0,0 +1,100 @@
+        if (200 !== $response->getStatusCode()) {
+            $error = $response->toArray(false);
+
+            throw new TransportException("Unable to send the SMS: {$error['title']}.", $response);
```
I didn't see this in my first review, but I would use sprintf here as well.

krasilnikovs replied:
Done and tests added

jschaedl reviewed (commented):
@krasilnikovm Thank you for your adjustments. Code looks good to me now, except the one line where `sprintf` could be used as well.

I know some of the bridges got merged without tests, but I think having tests here would be really great. Do you think you could add some?

krasilnikovs commented:
> 
> 
> @krasilnikovm Thank you for your adjustments. Code looks good to me now, except the one line where `sprintf` could be used as well.
> 
> I know some of the bridges got merged without tests, but I think having tests here would be really great. Do you think you could add some?

Hi @jschaedl, sorry for I'm not checked other places for using sprintf , and now I'm working on writing tests but I have noticed that SmsMessage::getOptions method always returns null, https://github.com/symfony/symfony/blob/0137609eaf04935a7762952ad6c8f26a87e32786/src/Symfony/Component/Notifier/Message/SmsMessage.php#L98-L101 
https://github.com/symfony/symfony/blob/0137609eaf04935a7762952ad6c8f26a87e32786/src/Symfony/Component/Notifier/Message/SmsMessage.php#L28-L37
 Can I add initializing options in the PR or better to create separate PR?

---------------------------------------------------------------------
Updated
I have added initializing options in the PR

stof commented on src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php:
```diff
@@ -0,0 +1,105 @@
+
+    public function __toString(): string
+    {
+        return sprintf('telnyx://%s?from=%s', $this->apiKey, $this->from);
```
@fabpot should `__toString` really expose the API key ?

stof replied:
btw, this does not look like it produces a valid DSN for the corresponding factory.

krasilnikovs replied:
My bad, Fixed

stof commented on src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json:
```diff
@@ -0,0 +1,32 @@
+    ],
+    "require": {
+        "php": ">=7.2.5",
+        "symfony/http-client": "^4.3|^5.0",
```
I would use `4.4` as min version rather than `4.3` as 4.3 is already EOL

krasilnikovs replied:
Fixed

fabpot commented on src/Symfony/Component/Notifier/Bridge/Telnyx/.gitattributes:
```diff
@@ -0,0 +1,3 @@
+/Tests export-ignore
+/phpunit.xml.dist export-ignore
+/.gitignore export-ignore
```
Missing `/.gitattributes export-ignore`

krasilnikovs replied:
Added

fabpot commented on src/Symfony/Component/Notifier/Bridge/Telnyx/README.md:
```diff
@@ -0,0 +1,12 @@
+Telnyx Notifier
+===============
+
+Provides Telnyx integration for Symfony Notifier.
```
Can you add a note about the DSN format (like done by most bridges)?

krasilnikovs replied:
Added

fabpot commented on src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php:
```diff
@@ -0,0 +1,114 @@
+{
+    /**
+     * @var array
+     */
```
Can be removed

krasilnikovs replied:
Removed

fabpot commented on src/Symfony/Component/Notifier/Message/SmsMessage.php:
```diff
@@ -25,15 +25,17 @@ final class SmsMessage implements MessageInterface
+    private $options;
 
-    public function __construct(string $phone, string $subject)
+    public function __construct(string $phone, string $subject, MessageOptionsInterface $options = null)
```
This change should be removed and discussed in another PR.

krasilnikovs replied:
PR created: https://github.com/symfony/symfony/pull/39365

fabpot reviewed (changes requested):
Looks like you forgot to update the `src/Symfony/Component/Notifier/Exception/UnsupportedSchemeException.php` file.

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json:
```diff
@@ -0,0 +1,32 @@
+    "require": {
+        "php": ">=7.2.5",
+        "symfony/http-client": "^4.4|^5.0",
+        "symfony/notifier": "^5.2"
```
Wouldn't it work with notifier 5.1 too?

Not sure how we handle this in general? 

cc @jschaedl @fabpot

derrabus replied:
If this PR is merged, it will be part of the Symfony 5.3 release. Symfony 5.1 will be EOL by then. I wouldn't change this line.

fabpot replied:
In any case, the code is not compatible with 5.1.

nicolas-grekas replied:
should use `~5.3.0`

krasilnikovs replied:
Done

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/.gitattributes:
```diff
@@ -0,0 +1,3 @@
+/Tests export-ignore
+/phpunit.xml.dist export-ignore
+/.gitignore export-ignore
```
```suggestion
/.gitattributes export-ignore
/.gitignore export-ignore
```

krasilnikovs replied:
Added

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/README.md:
```diff
@@ -0,0 +1,12 @@
+Telnyx Notifier
+===============
+
+Provides Telnyx integration for Symfony Notifier.
```
```suggestion
Provides [Telnyx](https://telnyx.com) integration for Symfony Notifier.
```

krasilnikovs replied:
Added

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php:
```diff
@@ -0,0 +1,114 @@
+ */
+final class TelnyxOptions implements MessageOptionsInterface
+{
+    /**
+     * @var array
+     */
```
```suggestion
```

krasilnikovs replied:
Removed

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportTest.php:
```diff
@@ -0,0 +1,276 @@
+
+    public function testShouldThrowExceptionWhenPassedNotSmsMessageInstance(): void
+    {
+        $this->expectException(LogicException::class);
+
+        $transport = $this->createTelnyxTransport();
+        $chatMessage = new ChatMessage('Hello');
```
```suggestion
        $transport = $this->createTelnyxTransport();
        $chatMessage = new ChatMessage('Hello');
        
        $this->expectException(LogicException::class);        
        
```

krasilnikovs replied:
Done

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportFactoryTest.php:
```diff
@@ -0,0 +1,53 @@
+ */
+final class TelnyxTransportFactoryTest extends TestCase
+{
+    public function testShouldCreateWithDsn(): void
```
please remove `:void` from all test classes

krasilnikovs replied:
Done

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php:
```diff
@@ -0,0 +1,48 @@
+        $scheme = $dsn->getScheme();
+        $apiKey = $dsn->getUser();
+        $host = 'default' === $dsn->getHost() ? null : $dsn->getHost();
+        $from = $dsn->getOption('from');
```
```suggestion
        $from = $dsn->getRequiredOption('from');        
```

It looks like `from` is required

krasilnikovs replied:
Done

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php:
```diff
@@ -0,0 +1,105 @@
+    protected function doSend(MessageInterface $message): SentMessage
+    {
+        if (!$message instanceof SmsMessage) {
+            throw new LogicException(sprintf('The "%s" transport only supports instances of "%s" (instance of "%s" given).', __CLASS__, SmsMessage::class, get_debug_type($message)));
```
Please use newly introduced `UnsupportedMessageTypeException` here.

Fore reference: https://github.com/symfony/symfony/pull/39365

Thanks

krasilnikovs replied:
Done

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php:
```diff
@@ -0,0 +1,48 @@
+     */
+    public function create(Dsn $dsn): TransportInterface
+    {
+        $scheme = $dsn->getScheme();
```
Please pull up the scheme check like here, to first check if the scheme is supported and afterwards for missing options. Thanks

https://github.com/symfony/symfony/blob/7d846d3e82f4e3dabec7d4b5300f50a1dd9ed7d6/src/Symfony/Component/Notifier/Bridge/Sinch/SinchTransportFactory.php#L27-L47

krasilnikovs replied:
Done

OskarStark commented:
Hi @krasilnikovm  👋 

Are you interested in finishing this PR or am I allowed to take over?

Cheers Oskar

krasilnikovs commented:
> 
> 
> Hi @krasilnikovm 👋
> 
> Are you interested in finishing this PR or am I allowed to take over?
> 
> Cheers Oskar

Hi, last month I was sick, now I feel nice and I'm going to finish the pr on weekend.

OskarStark commented:
> Hi, last month I was sick, now I feel nice and I'm going to finish the pr on weekend.

Glad to hear that  👍 😃

krasilnikovs commented:
> 
> 
> Looks like you forgot to update the `src/Symfony/Component/Notifier/Exception/UnsupportedSchemeException.php` file.

@fabpot Actually, I updated the file, or Am I something missed?
https://github.com/symfony/symfony/blob/1bbb34b12e5d484ce6cde5165a7ef978a38adc29/src/Symfony/Component/Notifier/Exception/UnsupportedSchemeException.php#L89-L95

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/LICENSE:
```diff
@@ -0,0 +1,19 @@
+Copyright (c) 2020 Fabien Potencier
```
```suggestion
Copyright (c) 2021 Fabien Potencier
```

krasilnikovs replied:
Done

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/README.md:
````diff
@@ -0,0 +1,23 @@
+DSN example
+-----------
+
+```
+// .env file
+TELNYX_DSN=telnyx://API_KEY@default?from=FROM_PHONE_NUMBER
+```
+
+where:
+ - `API_KEY` is your telnyx api key
+ - `FROM_PHONE_NUMBER` is your telnyx phone number
````
```suggestion
```
TELNYX_DSN=telnyx://API_KEY@default?from=FROM
```

where:
 - `API_KEY` is your telnyx api key
 - `FROM` is your telnyx phone number
```

krasilnikovs replied:
Done

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxOptions.php:
```diff
@@ -0,0 +1,111 @@
+
+/**
+ * @author Mihail Krasilnikov <mihail.krasilnikov.j@gmail.com>
+ *
+ * @experimental in 5.3
```
```suggestion
```

krasilnikovs replied:
done

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransport.php:
```diff
@@ -0,0 +1,106 @@
+
+/**
+ * @author Mihail Krasilnikov <mihail.krasilnikov.j@gmail.com>
+ *
+ * @experimental in 5.3
```
```suggestion
```

krasilnikovs replied:
done

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php:
```diff
@@ -0,0 +1,55 @@
+
+/**
+ * @author Mihail Krasilnikov <mihail.krasilnikov.j@gmail.com>
+ *
+ * @experimental in 5.3
```
```suggestion
```

krasilnikovs replied:
done

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/TelnyxTransportFactory.php:
```diff
@@ -0,0 +1,55 @@
+            throw new UnsupportedSchemeException($dsn, 'telnyx', $this->getSupportedSchemes());
+        }
+
+        $apiKey = $dsn->getUser();
+        $from = $dsn->getRequiredOption('from');
+
+        if (null === $apiKey) {
+            throw new IncompleteDsnException('Missing api key.'.$dsn->getOriginalDsn());
+        }
```
```suggestion
        $apiKey = $this->getUser($dsn);
        $from = $dsn->getRequiredOption('from');

```

krasilnikovs replied:
done

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportFactoryTest.php:
```diff
@@ -0,0 +1,53 @@
+
+/**
+ * @author Mihail Krasilnikov <mihail.krasilnikov.j@gmail.com>
+ *
+ * @experimental in 5.3
```
```suggestion
```

krasilnikovs replied:
done

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/Tests/TelnyxTransportTest.php:
```diff
@@ -0,0 +1,178 @@
+
+/**
+ * @author Mihail Krasilnikov <mihail.krasilnikov.j@gmail.com>
+ *
+ * @experimental in 5.3
```
```suggestion
```

krasilnikovs replied:
done

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json:
```diff
@@ -0,0 +1,32 @@
+    "license": "MIT",
+    "authors": [
+        {
+            "name": "Fabien Potencier",
+            "email": "fabien@symfony.com"
+        },
```
You should add yourself as an author here

krasilnikovs replied:
Added

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/composer.json:
```diff
@@ -0,0 +1,32 @@
+        "symfony/notifier": "^5.3"
+    },
+    "autoload": {
+        "psr-4": {
+            "Symfony\\Component\\Notifier\\Bridge\\Telnyx\\": ""
+        },
```
```suggestion
        "psr-4": { "Symfony\\Component\\Notifier\\Bridge\\Telnyx\\": "" },
```

krasilnikovs replied:
Done

OskarStark reviewed (approved):
Ready to merge after my comments 👍

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/CHANGELOG.md:
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+5.3.0
+-----
+
+ * Added the bridge
```
```suggestion
5.3
---

 * Add the bridge
```

krasilnikovs replied:
Done

OskarStark reviewed (changes requested):
This PR is blocked until https://github.com/symfony/symfony/pull/39774 is merged!

OskarStark commented on src/Symfony/Component/Notifier/Bridge/Telnyx/README.md:
````diff
@@ -0,0 +1,23 @@
+-----------
+
+```
+// .env file
````
```suggestion
```

krasilnikovs replied:
Done

OskarStark commented:
Can you please rebase to get rid of the conflicts? Thanks

OskarStark commented:
@krasilnikovm I need to close this PR, because it uses your other PR [[Notifier] Added ability to initialize sms options](https://github.com/symfony/symfony/pull/39774) which was closed ❌  without merging. 😢 

Feel free to reopen and update this PR or reopen a new one if you found another solution without the need of `MessageOptions` for SMS.

Thanks for your patience 🚀
